// react-md/lib/Sliders/Slider.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'react-dom', 'classnames', 'react-prop-types/lib/deprecated', 'react-prop-types/lib/isRequiredForA11y', '../constants/keyCodes', '../utils/getField', '../utils/EventUtils/isValidClick', '../utils/NumberUtils/calculateValueDistance', '../utils/NumberUtils/isWithinStep', '../utils/PropTypes/controlled', './SliderLabel', './Track', '../TextFields/TextField'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-dom'), require('classnames'), require('react-prop-types/lib/deprecated'), require('react-prop-types/lib/isRequiredForA11y'), require('../constants/keyCodes'), require('../utils/getField'), require('../utils/EventUtils/isValidClick'), require('../utils/NumberUtils/calculateValueDistance'), require('../utils/NumberUtils/isWithinStep'), require('../utils/PropTypes/controlled'), require('./SliderLabel'), require('./Track'), require('../TextFields/TextField'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactDom, global.classnames, global.deprecated, global.isRequiredForA11y, global.keyCodes, global.getField, global.isValidClick, global.calculateValueDistance, global.isWithinStep, global.controlled, global.SliderLabel, global.Track, global.TextField);
    global.Slider = mod.exports;
  }
})(this, function (exports, _react, _reactDom, _classnames, _deprecated, _isRequiredForA11y, _keyCodes, _getField, _isValidClick, _calculateValueDistance, _isWithinStep, _controlled, _SliderLabel, _Track, _TextField) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _deprecated2 = _interopRequireDefault(_deprecated);

  var _isRequiredForA11y2 = _interopRequireDefault(_isRequiredForA11y);

  var _getField2 = _interopRequireDefault(_getField);

  var _isValidClick2 = _interopRequireDefault(_isValidClick);

  var _calculateValueDistance2 = _interopRequireDefault(_calculateValueDistance);

  var _isWithinStep2 = _interopRequireDefault(_isWithinStep);

  var _controlled2 = _interopRequireDefault(_controlled);

  var _SliderLabel2 = _interopRequireDefault(_SliderLabel);

  var _Track2 = _interopRequireDefault(_Track);

  var _TextField2 = _interopRequireDefault(_TextField);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  var _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  function _objectWithoutProperties(obj, keys) {
    var target = {};

    for (var i in obj) {
      if (keys.indexOf(i) >= 0) continue;
      if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
      target[i] = obj[i];
    }

    return target;
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var Slider = function (_PureComponent) {
    _inherits(Slider, _PureComponent);

    function Slider(props) {
      _classCallCheck(this, Slider);

      var _this = _possibleConstructorReturn(this, (Slider.__proto__ || Object.getPrototypeOf(Slider)).call(this, props));

      var scale = Math.abs(props.max - props.min) / props.step;

      var value = typeof props.value !== 'undefined' ? props.value : props.defaultValue;

      var distance = value;
      if (distance === props.min) {
        distance = 0;
      } else if (distance === props.max) {
        distance = 100;
      } else if (props.min > 1 && props.defaultValue !== 0) {
        distance = Math.max(0, Math.min(100, (props.defaultValue - props.min) / (props.max - props.min) * 100));
      } else {
        distance = value / props.max * 100;
      }

      var thumbLeft = _this._calcLeft(distance);
      var trackFillWidth = distance + '%';

      if (typeof props.value !== 'undefined') {
        value = undefined;
      }

      var trackWidth = void 0;
      if (props.label && !props.editable && !props.leftIcon && !props.rightIcon) {
        trackWidth = '100%';
      }

      _this.state = {
        value: value,
        scale: scale,
        distance: distance,
        thumbLeft: thumbLeft,
        trackWidth: trackWidth,
        trackFillWidth: trackFillWidth,
        active: false,
        dragging: false,
        maskInked: false
      };
      _this._dragAdded = false;

      _this._focusThumb = _this._focusThumb.bind(_this);
      _this._updatePosition = _this._updatePosition.bind(_this);
      _this._handleDragStart = _this._handleDragStart.bind(_this);
      _this._handleDragMove = _this._handleDragMove.bind(_this);
      _this._handleDragEnd = _this._handleDragEnd.bind(_this);
      _this._handleFocus = _this._handleFocus.bind(_this);
      _this._handleKeyUp = _this._handleKeyUp.bind(_this);
      _this._handleKeyDown = _this._handleKeyDown.bind(_this);
      _this._handleIncrement = _this._handleIncrement.bind(_this);
      _this._handleTextFieldChange = _this._handleTextFieldChange.bind(_this);
      _this._blurOnOutsideClick = _this._blurOnOutsideClick.bind(_this);
      _this._calcTrackWidth = _this._calcTrackWidth.bind(_this);
      _this._animateDiscreteInk = _this._animateDiscreteInk.bind(_this);
      _this._setTrack = _this._setTrack.bind(_this);
      _this._setField = _this._setField.bind(_this);
      _this._setNode = _this._setNode.bind(_this);
      return _this;
    }

    _createClass(Slider, [{
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(nextProps) {
        var _props = this.props,
            leftIcon = _props.leftIcon,
            rightIcon = _props.rightIcon,
            label = _props.label,
            min = _props.min,
            max = _props.max,
            step = _props.step;

        if (leftIcon !== nextProps.leftIcon || rightIcon !== nextProps.rightIcon || label !== nextProps.label) {
          this._calcTrackWidth(nextProps);
        }

        if (min !== nextProps.min || max !== nextProps.max || step !== nextProps.step) {
          this.setState({ scale: Math.abs(nextProps.max - nextProps.min) / nextProps.step });
        }
      }
    }, {
      key: 'componentDidUpdate',
      value: function componentDidUpdate(prevProps, prevState) {
        var _state = this.state,
            active = _state.active,
            manualIncrement = _state.manualIncrement;

        var fn = window[(active ? 'add' : 'remove') + 'EventListener'];
        if (active !== prevState.active) {
          fn('click', this._blurOnOutsideClick);

          if (active) {
            this._focusThumb();
          }
        }

        var addDrag = active && !manualIncrement;
        fn = window[(addDrag ? 'add' : 'remove') + 'EventListener'];
        if (this._dragAdded !== addDrag) {
          fn('mousemove', this._handleDragMove);
          fn('mouseup', this._handleDragEnd);
          fn('touchmove', this._handleDragMove);
          fn('touchend', this._handleDragEnd);

          this._dragAdded = addDrag;
        }
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (this._dragAdded) {
          var rm = window.removeEventListener;
          rm('mousemove', this._handleMouseMove);
          rm('mouseup', this._handleMouseUp);
          rm('touchmove', this._handleDragMove);
          rm('touchend', this._handleDragEnd);
          rm('click', this._blurOnOutsideClick);
        }

        if (this._inkTimeout) {
          clearTimeout(this._inkTimeout);
        }

        if (this._focusTimeout) {
          clearTimeout(this._focusTimeout);
        }
      }
    }, {
      key: '_calcLeft',
      value: function _calcLeft(value) {
        return 'calc(' + value + '% - 6px)';
      }
    }, {
      key: '_isTextField',
      value: function _isTextField(target) {
        return this._textField && this._textField.contains(target);
      }
    }, {
      key: '_isValidClassList',
      value: function _isValidClassList(classList) {
        var invalid = false;
        ['md-slider-label', 'md-slider-ind', 'md-icon'].some(function (cl) {
          invalid = classList.contains(cl);
          return invalid;
        });

        return !invalid;
      }
    }, {
      key: '_updatePosition',
      value: function _updatePosition(e, normalize) {
        var x = (e.changedTouches ? e.changedTouches[0] : e).clientX;
        var scale = this.state.scale;
        var _props2 = this.props,
            onChange = _props2.onChange,
            onDragChange = _props2.onDragChange,
            min = _props2.min,
            max = _props2.max,
            step = _props2.step;

        var _calculateValueDistan = (0, _calculateValueDistance2.default)(x, this._track.offsetWidth, this._track.getBoundingClientRect().left, scale, step, min, max, normalize),
            value = _calculateValueDistan.value,
            distance = _calculateValueDistan.distance;

        if (onChange) {
          onChange(value, e);
        }

        if (!normalize && onDragChange) {
          onDragChange(distance, value, e);
        }

        var state = {
          value: value,
          active: true,
          distance: distance,
          manualIncrement: false,
          dragging: !normalize,
          thumbLeft: this._calcLeft(distance),
          trackFillWidth: distance + '%'
        };

        if (e.type === 'touchend' || e.type === 'mousedown') {
          state.maskInked = false;
        }

        if (typeof this.props.value !== 'undefined') {
          delete state.value;
        }

        this.setState(state);
      }
    }, {
      key: '_handleDragStart',
      value: function _handleDragStart(e) {
        if (e.type === 'mousedown' && this.props.onMouseDown) {
          this.props.onMouseDown(e);
        } else if (e.type === 'touchstart' && this.props.onTouchStart) {
          this.props.onTouchStart(e);
        }

        if (this.props.disabled || e.type === 'mousedown' && !(0, _isValidClick2.default)(e, 'mousedown') || e.target.className.match(/text-field/)) {
          return;
        }

        var classList = e.target.classList;

        var isDiscreteValue = classList.contains('md-slider-discrete-value');
        if (classList.contains('md-slider-thumb') || isDiscreteValue) {
          // Prevents text highlighting while dragging.
          e.preventDefault();
          this.setState({ dragging: true, active: true, manualIncrement: false, maskInked: false });
        } else if (!this._isTextField(e.target) && this._isValidClassList(classList)) {
          this._updatePosition(e, true);
        }
      }
    }, {
      key: '_setNode',
      value: function _setNode(node) {
        this._node = (0, _reactDom.findDOMNode)(node);
      }
    }, {
      key: '_setTrack',
      value: function _setTrack(track) {
        this._track = (0, _reactDom.findDOMNode)(track);
      }
    }, {
      key: '_setField',
      value: function _setField(field) {
        this._field = (0, _reactDom.findDOMNode)(field);
        this._calcTrackWidth(this.props);
      }
    }, {
      key: '_handleDragMove',
      value: function _handleDragMove(e) {
        if (this.props.disabled || !this.state.dragging) {
          return;
        }

        // Stops the text highlighting while dragging
        e.preventDefault();

        this._updatePosition(e, false);
      }
    }, {
      key: '_handleDragEnd',
      value: function _handleDragEnd(e) {
        if (!this.state.dragging || this.props.disabled || e.type === 'mouseup' && !(0, _isValidClick2.default)(e)) {
          return;
        }

        this._updatePosition(e, true);
      }
    }, {
      key: '_blurOnOutsideClick',
      value: function _blurOnOutsideClick(e) {
        if (this.state.dragging && !this.state.manualIncrement || this.props.disabled) {
          return;
        }

        if (!this._node.contains(e.target)) {
          this.setState({ active: false, maskInked: false });
        }
      }
    }, {
      key: '_handleIncrement',
      value: function _handleIncrement(incrementedValue, e, disableTransition) {
        var _props3 = this.props,
            onChange = _props3.onChange,
            min = _props3.min,
            max = _props3.max,
            discrete = _props3.discrete;


        var distance = Math.max(0, Math.min(100, (incrementedValue - min) / (max - min) * 100));
        var value = Math.max(min, Math.min(max, incrementedValue));

        if (onChange) {
          onChange(value, e);
        }

        var state = {
          distance: distance,
          manualIncrement: true,
          thumbLeft: this._calcLeft(distance),
          trackFillWidth: distance + '%',
          dragging: Math.abs(this.state.distance - distance) < 2 && disableTransition
        };

        if (typeof this.props.value === 'undefined') {
          state.value = value;
        }

        if (e.type === 'keydown' && !discrete) {
          state.maskInked = true;
        }

        this.setState(state);
      }
    }, {
      key: '_handleTextFieldChange',
      value: function _handleTextFieldChange(newValue, e) {
        this._handleIncrement(newValue, e, false);
      }
    }, {
      key: '_handleKeyDown',
      value: function _handleKeyDown(e) {
        var key = e.which || e.keyCode;
        var _props4 = this.props,
            min = _props4.min,
            max = _props4.max,
            step = _props4.step,
            disabled = _props4.disabled;

        if (disabled) {
          return;
        }

        if (key === _keyCodes.TAB) {
          this.setState({ active: false, maskInked: false });
          return;
        } else if (key !== _keyCodes.LEFT && key !== _keyCodes.RIGHT) {
          return;
        }

        var nextValue = (0, _getField2.default)(this.props, this.state);
        nextValue = Math.max(min, Math.min((key === _keyCodes.LEFT ? -step : step) + nextValue, max));

        this._handleIncrement(nextValue, e, true);
      }
    }, {
      key: '_handleKeyUp',
      value: function _handleKeyUp(e) {
        if ((e.which || e.keyCode) !== _keyCodes.TAB) {
          return;
        }

        if (this.props.discrete) {
          this._animateDiscreteInk();
        }

        this.setState({ maskInked: true });
      }
    }, {
      key: '_handleFocus',
      value: function _handleFocus() {
        this.setState({ active: true });
      }
    }, {
      key: '_calcTrackWidth',
      value: function _calcTrackWidth(props) {
        var editable = props.editable,
            leftIcon = props.leftIcon,
            rightIcon = props.rightIcon,
            inputWidth = props.inputWidth,
            label = props.label;


        if (!label) {
          this.setState({ trackWidth: null });
          return;
        }

        var trackWidth = null;
        if (!leftIcon && !rightIcon && !editable) {
          trackWidth = '100%';
        } else if (editable && !leftIcon) {
          var cs = window.getComputedStyle(this._textField);
          var pl = parseInt(cs.getPropertyValue('padding-left'), 10);
          var ml = parseInt(cs.getPropertyValue('margin-left'), 10);

          trackWidth = pl + ml + inputWidth;
        }

        if (trackWidth) {
          this.setState({ trackWidth: trackWidth });
        }
      }
    }, {
      key: '_animateDiscreteInk',
      value: function _animateDiscreteInk() {
        var _this2 = this;

        var wait = this.props.discreteInkTransitionTime;
        if (this._inkTimeout) {
          clearTimeout(this._inkTimeout);
        }

        this._inkTimeout = setTimeout(function () {
          _this2.setState({ leaving: true, maskInked: false });

          _this2._inkTimeout = setTimeout(function () {
            _this2._inkTimeout = null;
            _this2.setState({ leaving: false });
          }, wait);
        }, wait);
      }
    }, {
      key: '_focusThumb',
      value: function _focusThumb() {
        var _this3 = this;

        if (this._focusTimeout) {
          clearTimeout(this._focusTimeout);
        }

        this._focusTimeout = setTimeout(function () {
          _this3._focusTimeout = null;
          if (!_this3._thumb) {
            _this3._thumb = _this3._node.querySelector('.md-slider-thumb');
          }

          _this3._thumb.focus();
        }, 100);
      }
    }, {
      key: 'render',
      value: function render() {
        var _state2 = this.state,
            dragging = _state2.dragging,
            active = _state2.active,
            thumbLeft = _state2.thumbLeft,
            trackFillWidth = _state2.trackFillWidth,
            maskInked = _state2.maskInked,
            trackWidth = _state2.trackWidth,
            scale = _state2.scale,
            distance = _state2.distance;

        var _props5 = this.props,
            id = _props5.id,
            min = _props5.min,
            max = _props5.max,
            disabled = _props5.disabled,
            className = _props5.className,
            trackStyle = _props5.trackStyle,
            trackClassName = _props5.trackClassName,
            thumbStyle = _props5.thumbStyle,
            thumbClassName = _props5.thumbClassName,
            discreteValueStyle = _props5.discreteValueStyle,
            discreteValueClassName = _props5.discreteValueClassName,
            label = _props5.label,
            editable = _props5.editable,
            step = _props5.step,
            inputWidth = _props5.inputWidth,
            leftIcon = _props5.leftIcon,
            rightIcon = _props5.rightIcon,
            discrete = _props5.discrete,
            discreteTicks = _props5.discreteTicks,
            tickWidth = _props5.tickWidth,
            valuePrecision = _props5.valuePrecision,
            props = _objectWithoutProperties(_props5, ['id', 'min', 'max', 'disabled', 'className', 'trackStyle', 'trackClassName', 'thumbStyle', 'thumbClassName', 'discreteValueStyle', 'discreteValueClassName', 'label', 'editable', 'step', 'inputWidth', 'leftIcon', 'rightIcon', 'discrete', 'discreteTicks', 'tickWidth', 'valuePrecision']);

        delete props.value;
        delete props.onChange;
        delete props.onDragChange;
        delete props.discreteInkTransitionTime;

        // delete deprecated
        delete props.stepPrecision;

        var value = (0, _getField2.default)(this.props, this.state);
        var rightChildren = rightIcon;
        if (editable) {
          rightChildren = _react2.default.createElement(_TextField2.default, {
            id: id + 'Editor',
            ref: this._setField,
            type: 'number',
            value: value,
            inputClassName: 'md-slider-editor',
            style: { width: inputWidth },
            onChange: this._handleTextFieldChange,
            step: step
          });
        }

        return _react2.default.createElement(
          'div',
          _extends({}, props, {
            ref: this._setNode,
            className: (0, _classnames2.default)('md-slider-container', className, {
              'md-pointer--hover': !disabled
            }),
            onMouseDown: this._handleDragStart,
            onTouchStart: this._handleDragStart
          }),
          _react2.default.createElement(
            _SliderLabel2.default,
            { htmlFor: id },
            label
          ),
          _react2.default.createElement('input', {
            id: id,
            type: 'range',
            className: 'md-slider-input',
            readOnly: true,
            min: min,
            max: max,
            value: value,
            disabled: disabled
          }),
          leftIcon,
          _react2.default.createElement(_Track2.default, {
            ref: this._setTrack,
            style: Object.assign({}, trackStyle, { width: trackWidth }),
            className: (0, _classnames2.default)(trackClassName, {
              'md-slider-track--ind-left': leftIcon,
              'md-slider-track--ind-right': rightIcon
            }),
            thumbStyle: thumbStyle,
            thumbClassName: thumbClassName,
            discreteValueStyle: discreteValueStyle,
            discreteValueClassName: discreteValueClassName,
            active: active,
            dragging: dragging,
            disabled: disabled,
            thumbLeft: thumbLeft,
            trackFillWidth: trackFillWidth,
            on: !disabled && distance > min,
            off: distance === 0,
            maskInked: maskInked,
            onThumbKeyUp: this._handleKeyUp,
            onThumbKeyDown: this._handleKeyDown,
            onThumbFocus: this._handleFocus,
            discrete: discrete,
            tickWidth: tickWidth,
            discreteTicks: discreteTicks,
            valuePrecision: valuePrecision,
            step: step,
            scale: scale,
            value: value
          }),
          rightChildren
        );
      }
    }]);

    return Slider;
  }(_react.PureComponent);

  Slider.propTypes = {
    /**
     * An id to use for the `Slider`. This is required if the `label` prop
     * is defined.
     */
    id: function id(props, propName, component) {
      for (var _len = arguments.length, others = Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
        others[_key - 3] = arguments[_key];
      }

      if (typeof props.label === 'undefined') {
        return _react.PropTypes.string.apply(_react.PropTypes, [props, propName, component].concat(others));
      }

      return (0, _isRequiredForA11y2.default)(_react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string])).apply(undefined, [props, propName, component].concat(others));
    },

    /**
     * An optional style to apply to the slider's container.
     */
    style: _react.PropTypes.object,

    /**
     * An optional className to apply to the slider's container.
     */
    className: _react.PropTypes.string,

    /**
     * An optional style to apply to the slider's thumb.
     */
    thumbStyle: _react.PropTypes.object,

    /**
     * An optionl className to apply to the slider's thumb.
     */
    thumbClassName: _react.PropTypes.string,

    /**
     * An optional style to apply to the slider's track.
     */
    trackStyle: _react.PropTypes.object,

    /**
     * An optional className to apply to the slider's track.
     */
    trackClassName: _react.PropTypes.string,

    /**
     * An optional style to apply to the slider's track fill.
     */
    trackFillStyle: _react.PropTypes.object,

    /**
     * An optional className to apply to the slider's track fill.
     */
    trackFillClassName: _react.PropTypes.string,

    /**
     * An optional style to apply to a discrete slider's value.
     */
    discreteValueStyle: _react.PropTypes.object,

    /**
     * An optional className to apply to a discrete slider's value.
     */
    discreteValueClassName: _react.PropTypes.string,

    /**
     * The default value for the slider.
     */
    defaultValue: _react.PropTypes.number.isRequired,

    /**
     * The min value for the slider. The min and max values must be on the same
     * side of 0. It is currently invalid to have a range from negative to positive.
     */
    min: function min(props, propName, component) {
      for (var _len2 = arguments.length, others = Array(_len2 > 3 ? _len2 - 3 : 0), _key2 = 3; _key2 < _len2; _key2++) {
        others[_key2 - 3] = arguments[_key2];
      }

      var _PropTypes$number;

      var err = (_PropTypes$number = _react.PropTypes.number).isRequired.apply(_PropTypes$number, [props, propName, component].concat(others));
      if (!err) {
        var min = props[propName];
        var name = void 0;
        if (min > props.value) {
          name = 'value';
        } else if (min > props.defaultValue) {
          name = 'defaultValue';
        }

        if (name) {
          err = new Error('The \'' + propName + '\' prop must be less than or equal to the \'' + name + '\' prop for the \'' + component + '\' but ' + ('received: \'min: ' + min + '\' and \'' + name + ': ' + props[name] + '\''));
        } else if (min < 0 && props.max > 0) {
          err = new Error('The \'' + component + '\' is unable to have a range spanning from negative to positive. The range ' + 'must either be all negative or all postive.');
        }
      }

      return err;
    },

    /**
     * The max value for the slider. The max value must be greater than the min value
     * and the min and max values must be on the same side of 0. It is currently
     * invalid to have a range from negative to positive.
     */
    max: function max(props, propName, component) {
      for (var _len3 = arguments.length, others = Array(_len3 > 3 ? _len3 - 3 : 0), _key3 = 3; _key3 < _len3; _key3++) {
        others[_key3 - 3] = arguments[_key3];
      }

      var _PropTypes$number2;

      var err = (_PropTypes$number2 = _react.PropTypes.number).isRequired.apply(_PropTypes$number2, [props, propName, component].concat(others));
      if (!err) {
        var max = props[propName];
        var name = void 0;
        if (max < props.value) {
          name = 'value';
        } else if (max < props.defaultValue) {
          name = 'defaultValue';
        }

        if (name) {
          err = new Error('The \'' + propName + '\' prop must be greater than or equal to the \'' + name + '\' prop for the \'' + component + '\' but ' + ('received: \'' + propName + ': ' + max + '\' and \'' + name + ': ' + props[name] + '\''));
        }
      }

      return err;
    },

    /**
     * Boolean if the slider is disabled.
     */
    disabled: _react.PropTypes.bool,

    /**
     * An optional value for the slider. This will make the component controlled
     * and require the `onChange` function.
     */
    value: (0, _controlled2.default)(_react.PropTypes.number, 'onChange'),

    /**
     * This is called when the slider's value gets updated. The value can be updated
     * by one of the following:
     *
     * - Clicking a section of the slider
     * - Dragging the slider with the mouse
     * - Touching a section of the slider.
     * - Dragging the slider with touch
     * - Using the text field to update the value either by typing or incrementing
     * - Using the left or right arrow keys to increment/decrement the value.
     *
     * The callback for this function is as follows:
     *
     * ```js
     * onChange(value, event);
     * ```
     *
     * where the event can either be:
     * - a touch start event
     * - a touch move event
     * - a touch end event
     * - a mouse down event
     * - a mouse move event
     * - a mouse up event
     * - a key up event
     * - a key down event
     */
    onChange: _react.PropTypes.func,

    /**
     * This is only called when the user is dragging the slider with either
     * the mouse or touch. Probably not really usefull. It just includes the
     * new drag percentage while the `onChange` does not.
     *
     * The callback for this function is defined as:
     *
     * ```js
     * onDragChange(dragPercentage, value, (touchMove || mouseMove) event);
     * ```
     */
    onDragChange: _react.PropTypes.func,

    /**
     * An optional function to call when the slider's container has
     * the mousedown event.
     */
    onMouseDown: _react.PropTypes.func,

    /**
     * An optional function to call when the slider's container has
     * the touchstart event.
     */
    onTouchStart: _react.PropTypes.func,

    /**
     * An optional icon or letter to place to the left of the slider.
     * if you want to use a non-font icon or a letter, use the `md-slider-ind`
     * className on your element.
     */
    leftIcon: _react.PropTypes.element,

    /**
     * An optional icon or letter to place to the right of the slider.
     * if you want to use a non-font icon or a letter, use the `md-slider-ind`
     * className on your element.
     *
     * > NOTE: This can not be used if the `editable` prop is true.
     */
    rightIcon: _react.PropTypes.element,

    /**
     * An optional label to display above the slider. If this prop
     * is set, then an `id` must also be given.
     */
    label: _react.PropTypes.node,

    /**
     * The incremental amount when the user hits left or right with the
     * keyboard arrows, or the user hits the up or down buttons in the
     * editable number text field. This number must be a number between
     * 0 and 1 or a whole number above 1.
     */
    step: function step(props, propName, component) {
      for (var _len4 = arguments.length, others = Array(_len4 > 3 ? _len4 - 3 : 0), _key4 = 3; _key4 < _len4; _key4++) {
        others[_key4 - 3] = arguments[_key4];
      }

      var _PropTypes$number3;

      var err = (_PropTypes$number3 = _react.PropTypes.number).isRequired.apply(_PropTypes$number3, [props, propName, component].concat(others));
      if (!err) {
        var step = props[propName];
        if (step <= 0) {
          err = new Error('The \'' + propName + '\' for the \'' + component + '\' must be a number greater than 0. The ' + ('current value is \'' + step + '\'.'));
        } else {
          var valueDefined = typeof props.value !== 'undefined';
          var name = void 0;
          if (!(0, _isWithinStep2.default)(valueDefined ? props.value : props.defaultValue, step)) {
            name = valueDefined ? 'value' : 'defaultValue';
          }

          if (name) {
            err = new Error('The \'' + name + '\' prop on \'' + component + '\' should be a number divisible by the ' + ('\'' + propName + '\' prop. The current value is \'' + props[name] + '\' and the \'' + propName + '\' ') + ('is \'' + step + '\'.'));
          }
        }
      }

      return err;
    },

    /**
     * Boolean if the Slider should be editable. This will place a number text field
     * to the right of the slider. If this prop is set to `true`, the `rightIcon`
     * prop can not be set.
     */
    editable: function editable(props, propName, component) {
      for (var _len5 = arguments.length, others = Array(_len5 > 3 ? _len5 - 3 : 0), _key5 = 3; _key5 < _len5; _key5++) {
        others[_key5 - 3] = arguments[_key5];
      }

      var _PropTypes$bool;

      if (typeof props[propName] === 'undefined') {
        return null;
      }

      var err = (_PropTypes$bool = _react.PropTypes.bool).isRequired.apply(_PropTypes$bool, [props, propName, component].concat(others));
      if (!err && typeof props.rightIcon !== 'undefined') {
        err = new Error('The \'' + component + '\' is unable to be \'editable\' and include a \'rightIcon\'.');
      }

      return err;
    },

    /**
     * The width for the number text field when the Slider is editable.
     */
    inputWidth: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired,

    /**
     * Boolean if the slider should be discrete. This will update the slider to include a
     * _balloon_ with the current value inside. It will also not allow the `Slider` to be
     * editable.
     */
    discrete: function discrete(props, propName, component) {
      for (var _len6 = arguments.length, others = Array(_len6 > 3 ? _len6 - 3 : 0), _key6 = 3; _key6 < _len6; _key6++) {
        others[_key6 - 3] = arguments[_key6];
      }

      if (typeof props[propName] === 'undefined') {
        return null;
      }

      var err = _react.PropTypes.bool.apply(_react.PropTypes, [props, propName, component].concat(others));
      if (!err && typeof props.editable !== 'undefined') {
        err = new Error('The \'' + component + '\' cannot be \'discrete\' and \'editable\'. Please choose one.');
      }

      return err;
    },

    /**
     * The width of each tick for a discrete slider with ticks. This can either be a number
     * which gets converted to `px`, or a valid css unit.
     */
    tickWidth: _react.PropTypes.oneOfType([_react.PropTypes.number, _react.PropTypes.string]).isRequired,

    /**
     * This is number divisible by the total number of values included in the Slider. Every
     * value that is divisible by this number will include a tick mark. It is common recommended
     * to have this equal to the `step` prop.
     *
     * This prop is completely optional.
     */
    discreteTicks: function discreteTicks(props, propName, component) {
      for (var _len7 = arguments.length, others = Array(_len7 > 3 ? _len7 - 3 : 0), _key7 = 3; _key7 < _len7; _key7++) {
        others[_key7 - 3] = arguments[_key7];
      }

      if (typeof props[propName] === 'undefined') {
        return null;
      }

      var err = _react.PropTypes.number.apply(_react.PropTypes, [props, propName, component].concat(others));
      if (!err) {
        var min = props.min,
            max = props.max,
            step = props.step;

        var range = Math.abs(max - min);

        if (range / props[propName] % 1 !== 0) {
          err = new Error('The \'' + propName + '\' must be a number divisible by the range set by the \'min\' and ' + ('\'max\' props. The current range is \'' + range + '\' by including the min: \'' + min + '\' and ') + ('max: \'' + max + '\' values. The current value of \'' + propName + '\' is \'' + props[propName] + '\'.'));
        } else if (props[propName] % step !== 0) {
          err = new Error('The \'step\' prop must be a number divisible by the \'' + propName + '\'. It is common to have ' + ('them as the same value. The current \'step\' is \'' + step + '\' and the \'' + propName + '\' is \'' + props[propName] + '\'.'));
        }
      }

      return err;
    },

    /**
     * The transition time for a discrete Slider's keyboard focus ink. This should match the
     * `md-slider-discrete-ink-transition-time` value in your SCSS. This is used because
     * the ink is only visible temporarily for a discrete slider when keyboard focusing.
     */
    discreteInkTransitionTime: _react.PropTypes.number.isRequired,

    /**
     * The precision that the value should be rounded to when the Slider is updated. This
     * needs to be a whole number greater than or equal to 0.
     */
    valuePrecision: function valuePrecision(props, propName, component) {
      for (var _len8 = arguments.length, others = Array(_len8 > 3 ? _len8 - 3 : 0), _key8 = 3; _key8 < _len8; _key8++) {
        others[_key8 - 3] = arguments[_key8];
      }

      var _PropTypes$number4;

      var err = (_PropTypes$number4 = _react.PropTypes.number).isRequired.apply(_PropTypes$number4, [props, propName, component].concat(others));
      if (!err) {
        var precision = props[propName];

        if (precision % 1 !== 0 || precision < 0) {
          err = new Error('The \'' + propName + '\' must be a positive whole number or 0 on the \'' + component + '\'. ' + ('The current \'' + propName + '\' is \'' + precision + '\''));
        }
      }

      return err;
    },

    stepPrecision: (0, _deprecated2.default)(_react.PropTypes.number, 'Use `step` and `valuePreceision` instead')
  };
  Slider.defaultProps = {
    defaultValue: 0,
    min: 0,
    max: 100,
    step: 1,
    inputWidth: 40,
    tickWidth: 6,
    discreteInkTransitionTime: 300,
    valuePrecision: 0
  };
  exports.default = Slider;
});
