// react-md/lib/Tooltips/TooltipContainer.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(['exports', 'react', 'react-dom', 'react-addons-transition-group', 'classnames', '../constants/keyCodes', '../utils/EventUtils/captureNextEvent', './Tooltip'], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require('react'), require('react-dom'), require('react-addons-transition-group'), require('classnames'), require('../constants/keyCodes'), require('../utils/EventUtils/captureNextEvent'), require('./Tooltip'));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactDom, global.reactAddonsTransitionGroup, global.classnames, global.keyCodes, global.captureNextEvent, global.Tooltip);
    global.TooltipContainer = mod.exports;
  }
})(this, function (exports, _react, _reactDom, _reactAddonsTransitionGroup, _classnames, _keyCodes, _captureNextEvent, _Tooltip) {
  'use strict';

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _reactAddonsTransitionGroup2 = _interopRequireDefault(_reactAddonsTransitionGroup);

  var _classnames2 = _interopRequireDefault(_classnames);

  var _captureNextEvent2 = _interopRequireDefault(_captureNextEvent);

  var _Tooltip2 = _interopRequireDefault(_Tooltip);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
      throw new TypeError("Cannot call a class as a function");
    }
  }

  var _createClass = function () {
    function defineProperties(target, props) {
      for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
      }
    }

    return function (Constructor, protoProps, staticProps) {
      if (protoProps) defineProperties(Constructor.prototype, protoProps);
      if (staticProps) defineProperties(Constructor, staticProps);
      return Constructor;
    };
  }();

  function _possibleConstructorReturn(self, call) {
    if (!self) {
      throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }

    return call && (typeof call === "object" || typeof call === "function") ? call : self;
  }

  function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
      throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }

    subClass.prototype = Object.create(superClass && superClass.prototype, {
      constructor: {
        value: subClass,
        enumerable: false,
        writable: true,
        configurable: true
      }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
  }

  var TooltipContainer = function (_PureComponent) {
    _inherits(TooltipContainer, _PureComponent);

    function TooltipContainer(props) {
      _classCallCheck(this, TooltipContainer);

      var _this = _possibleConstructorReturn(this, (TooltipContainer.__proto__ || Object.getPrototypeOf(TooltipContainer)).call(this, props));

      _this.state = { visible: false };
      _this._delayedTimeout = null;

      _this._setContainers = _this._setContainers.bind(_this);
      _this._showTooltip = _this._showTooltip.bind(_this);
      _this._hideTooltip = _this._hideTooltip.bind(_this);
      _this._stopContextMenu = _this._stopContextMenu.bind(_this);
      _this._handleKeyUp = _this._handleKeyUp.bind(_this);
      return _this;
    }

    _createClass(TooltipContainer, [{
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (this._container) {
          this._container.removeEventListener('mouseover', this._showTooltip);
          this._container.removeEventListener('mouseleave', this._hideTooltip);
          this._container.removeEventListener('touchstart', this._showTooltip);
          this._container.removeEventListener('touchend', this._hideTooltip);
          this._container.removeEventListener('keyup', this._handleKeyUp);
          this._container.removeEventListener('blur', this._hideTooltip);
        }

        if (this._delayedTimeout) {
          clearTimeout(this._delayedTimeout);
        }
      }
    }, {
      key: '_setContainers',
      value: function _setContainers(container) {
        if (container) {
          this._container = (0, _reactDom.findDOMNode)(container).parentNode;

          this._container.addEventListener('mouseover', this._showTooltip);
          this._container.addEventListener('mouseleave', this._hideTooltip);
          this._container.addEventListener('touchstart', this._showTooltip);
          this._container.addEventListener('touchend', this._hideTooltip);
          this._container.addEventListener('keyup', this._handleKeyUp);
          this._container.addEventListener('blur', this._hideTooltip);
        }
      }
    }, {
      key: '_stopContextMenu',
      value: function _stopContextMenu(e) {
        e.preventDefault();
        window.removeEventListener('contextmenu', this._stopContextMenu, true);
        (0, _captureNextEvent2.default)('click');
        this.setState({ visible: true });
      }
    }, {
      key: '_showTooltip',
      value: function _showTooltip(e) {
        var _this2 = this;

        if (e.type === 'mouseover' && this._touched) {
          return;
        }

        if (e.type === 'touchstart') {
          this._touched = true;

          window.addEventListener('contextmenu', this._stopContextMenu, true);
          return;
        }

        var delay = this.props.delay;

        if (this._delayedTimeout) {
          clearTimeout(this._delayedTimeout);
        }

        if (delay) {
          this._delayedTimeout = setTimeout(function () {
            _this2._delayedTimeout = null;

            _this2.setState({ visible: true });
          }, delay);
        } else {
          this.setState({ visible: true });
        }
      }
    }, {
      key: '_hideTooltip',
      value: function _hideTooltip(e) {
        if (this._delayedTimeout) {
          clearTimeout(this._delayedTimeout);
        }

        if (e.type === 'mouseover' && this._touched) {
          return;
        }

        this.setState({ visible: false });
      }
    }, {
      key: '_handleKeyUp',
      value: function _handleKeyUp(e) {
        if ((e.which || e.keyCode) === _keyCodes.TAB) {
          this._showTooltip(e);
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var visible = this.state.visible;
        var _props = this.props,
            style = _props.style,
            className = _props.className,
            tooltipStyle = _props.tooltipStyle,
            tooltipClassName = _props.tooltipClassName,
            label = _props.label,
            position = _props.position,
            enterTimeout = _props.enterTimeout,
            leaveTimeout = _props.leaveTimeout;


        var tooltip = _react2.default.createElement(
          _Tooltip2.default,
          {
            key: 'tooltip',
            style: tooltipStyle,
            className: tooltipClassName,
            position: position,
            enterTimeout: enterTimeout,
            leaveTimeout: leaveTimeout
          },
          label
        );

        return _react2.default.createElement(
          _reactAddonsTransitionGroup2.default,
          {
            style: style,
            className: (0, _classnames2.default)('md-tooltip-container', className),
            ref: this._setContainers,
            component: 'div'
          },
          visible ? tooltip : null
        );
      }
    }]);

    return TooltipContainer;
  }(_react.PureComponent);

  TooltipContainer.propTypes = {
    style: _react.PropTypes.object,
    className: _react.PropTypes.string,
    tooltipStyle: _react.PropTypes.object,
    tooltipClassName: _react.PropTypes.string,
    label: _react.PropTypes.node.isRequired,
    position: _Tooltip2.default.propTypes.position,
    delay: _react.PropTypes.number.isRequired,
    enterTimeout: _react.PropTypes.number.isRequired,
    leaveTimeout: _react.PropTypes.number.isRequired
  };
  exports.default = TooltipContainer;
});
